### PR TITLE
Add support for virtual projects

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

The basic idea here is: any project can either be a package, or not ("virtual").

If a project is virtual, we don't build or install it.

A project is virtual if either of the following are true:

- `tool.uv.virtual = true` is set.
- `[build-system]` is absent.

The concept of "virtual projects" only applies to workspace member right now; it doesn't apply to `path` dependencies which are treated like arbitrary Python source trees.

TODOs that should be resolved prior to merging:

- [ ] Documentation
- [ ] How do we reconcile this with "virtual workspace roots" which are a little different -- they omit `[project]` entirely and don't even have a name?
- [x] `uv init --virtual` should create a virtual project rather than a virtual workspace.
- [x] Running `uv sync` in a virtual project after `uv init --virtual` shows `Audited 0 packages in 0.01ms`, which is awkward. (See: https://github.com/astral-sh/uv/pull/6588.)

Closes https://github.com/astral-sh/uv/issues/6511.

### CODE REVIEW

File: crates/uv/tests/sync.rs [hunk 1]
```diff
@@ -55,6 +59,10 @@ fn locked() -> Result<()> {
         version = "0.1.0"
         requires-python = ">=3.12"
         dependencies = ["anyio==3.7.0"]
+
+        [build-system]
+        requires = ["setuptools>=42", "wheel"]
+        build-backend = "setuptools.build_meta"
```
Comment (charliermarsh) [inline crates/uv/tests/sync.rs:65]: Unfortunately, we do need to set an explicit build backend for our test projects now. This is fine and correct, we were just omitting it before.

**Comment by charliermarsh** [comment]
This needs documentation.

**Comment by charliermarsh** [comment]
I wonder if we should use different terminology here to differentiate from a "virtual workspace root".

**Comment by charliermarsh** [comment]
We might be better off with a concept of "applications" vs. "libraries" here?

**Comment by charliermarsh** [comment]
I feel pretty good about this apart from the use of "virtual" here and the fact that a "virtual workspace root" is slightly different.

**Comment by mkniewallner** [comment]
"virtual" seems a bit abstract to me, and not really self-explanatory about what it does.

Some background on this:
- PDM uses a [`distribution`](https://pdm-project.org/en/latest/usage/project/#library-or-application) boolean
- Poetry uses a [`package-mode`](https://python-poetry.org/docs/basic-usage/#operating-modes) boolean

From those two, I find Poetry's one to be a better fit, as "distribution" seems like this is only about preventing publishing a package.

What about using something a bit more explicit like `project-type` that could either be `application` or `package` (or `library`)? I personally find those concepts closer to what people are used to when creating Python projects. Using an enum also make it possible to add more values in the future (although I don't really see what else we could have).

**Comment by charliermarsh** [comment]
My preferred framing here is application vs. library (in Rust it’s “binary” vs. “library”), but curious to hear from @zanieb on it when they’re back.

**Comment by a-reich** [comment]
IIUC, [project] still must have name and version for virtual projects/applications - does it make sense to make them optional?

**Comment by charliermarsh** [comment]
No, I still plan to require those. Otherwise, it’s not clear how to identify a given package in the lockfile.

**Comment by a-reich** [comment]
Hmm ok, I guess I misunderstood the mental model, as it hadn’t occurred to me that a project was allowed to depend on an application.

**Comment by charliermarsh** [comment]
Just depends on the semantics we want to allow!

**Comment by philgyford** [comment]
> No, I still plan to require those. Otherwise, it’s not clear how to identify a given package in the lockfile.

Fair enough if there’s no technical way around it, but it feels very odd to require it for something like a website.

Even on my solo, personal sites I might commit and deploy several times a day, and having to decide on a version number every time (and remember to change it) would be a real pain.

**Comment by z3z1ma** [comment]
> Even on my solo, personal sites I might commit and deploy several times a day, and having to decide on a version number every time (and remember to change it) would be a real pain.

You dont need to. Set the fields once in pyproject toml. Since you are not building a package with a build backend, the values are moot. Just use them as metadata for your app.

**Comment by sersorrel** [comment]
I'm not sure that the "library"/"application" framing necessarily corresponds well to "should/should not be installable". You want a library to be an installable package, so that it can be depended upon; but you also (sometimes?) want an application to be an installable package, so that you can install and run it (say, `pipx install uv`).

**Comment by minkooseo** [comment]
What about "non package" or "not a package" then? Not sure if it's a good flag name, but it's something we can use to explain what this is about. It's also a term used in #6578. 

Coming from the experience of always putting sources at the top, 'git push' and then 'git pull' at the server to serve web traffic, it was quite new to me that I should put code under src/project_name. I learned that I build a package and publish it that way, and I can create a package and install it on the server.

It is still a project for me, so I was searching the doc to understand what's a project then if mine is _virtual_. (I still don't know)

It is an application and not a library, but it is now clear to me that application can still be an installable package. It may create confusion for people with knowledge.

**Comment by charliermarsh** [comment]
That's actually what we're using in this PR. The flag is `package = false` etc.

**Comment by zanieb** [comment]
@minkooseo might be helpful to look at #6689 too, but we're going to push out a bunch of changes around this soon.

**Review by zanieb — APPROVED** [review]

**Pull request merged**